tfhe/src/integer/bigint/i256.rs

```diff
@@ -0,0 +1,152 @@
+    fn test_i128_conversion() {
+        let input = I256::from(-1i128);
+        assert_eq!(input.0, [u64::MAX, u64::MAX, u64::MAX, u64::MAX]);
+        assert_eq!(i128::cast_from(input), -1i128);
+
+        // 0b11111111...1000...00
+        // input is 64 0s in MSB, 64 1s in LSB
+        let input = i128::from(i64::MIN) << 1;
+        let big = I256::from(input);
+        assert_eq!(
+            big.0,
+            [(i64::MIN << 1) as u64, u64::MAX, u64::MAX, u64::MAX]
+        );
+        assert_eq!(i128::cast_from(big), input);
+    }
+
+    #[test]
+    fn test_add_wrap_around() {
+        assert_eq!(I256::MAX + I256::ONE, I256::MIN);
+        assert_eq!(I256::MIN + (-I256::ONE), I256::MAX);
+    }
+
+    #[test]
+    fn test_sub_wrap_around() {
+        assert_eq!(I256::MIN - I256::ONE, I256::MAX);
+        assert_eq!(I256::MAX - (-I256::ONE), I256::MIN);
+    }
+
+    #[test]
+    fn test_add() {
+        assert_eq!(I256::from(-1i32) + I256::ONE, I256::ZERO);
+    }
+
+    #[test]
+    fn test_sub() {
+        assert_eq!(I256::ONE - I256::ONE, I256::ZERO);
+        assert_eq!(I256::ONE - I256::from(-1i32), I256::TWO);
+    }
+
+    #[test]
+    fn test_shl_limits() {
+        assert_eq!(
+            I256::ONE << I256::BITS,
+            I256::ONE << (I256::BITS % I256::BITS)
+        );
+        assert_eq!(
+            I256::ONE << I256::BITS,
+            I256::ONE << (I256::BITS % I256::BITS)
+        );
+
+        // We aim to have same behaviour as rust native types
+        assert_eq!(1i128.wrapping_shl(128), 1i128 << (128 % i128::BITS));
+        assert_eq!(1i128.wrapping_shl(129), 1i128 << (129 % i128::BITS));
+    }
+
+    #[test]
+    fn test_shl() {
+        let input = i64::MIN as i128;
+        let a = I256::from(input);
+
+        assert_eq!(a << 1u32, I256::from(input << 1));
+    }
+
+    #[test]
+    fn test_div_rem() {
+        let i64_max = I256::from(i64::MAX);
+        let (expected_q, expected_r) = (i64::MAX / i64::MAX, i64::MAX % i64::MAX);
+        assert_eq!(i64_max / i64_max, I256::from(expected_q));
+        assert_eq!(i64_max % i64_max, I256::from(expected_r));
+
+        let mut rng = rand::thread_rng();
+        for _ in 0..5 {
+            let a = rng.gen::<i128>();
+            let b = rng.gen::<i128>();
+
+            let res_q = I256::from(a) / I256::from(b);
+            let res_r = I256::from(a) % I256::from(b);
+            let expected_q = a / b;
+            let expected_r = a % b;
+            assert_eq!(res_q, I256::from(expected_q));
+            assert_eq!(res_r, I256::from(expected_r));
+        }
+
+        let i256_max = I256::MAX;
+        let res_q = i256_max / I256::ONE;
+        let res_r = i256_max % I256::ONE;
+        assert_eq!(res_q, i256_max);
+        assert_eq!(res_r, I256::ZERO);
+        let a = I256::from((
+            330090607680070657243007942644838587352u128,
+            16697987949408356462265050424072830869u128,
+        ));
+        let a = !a + I256::ONE;
```

Comment:
maybe let minus_a = ... ? with a small comment saying the test value comes from a python example